pkg/agent/controller/networkpolicy/node_reconciler_other.go

```diff
@@ -0,0 +1,769 @@
+	"antrea.io/antrea/pkg/agent/types"
+	"antrea.io/antrea/pkg/agent/util/iptables"
+	"antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	secv1beta1 "antrea.io/antrea/pkg/apis/crd/v1beta1"
+	"antrea.io/antrea/pkg/util/ip"
+)
+
+const (
+	ipsetPrefix         = "ANTREA-POL"
+	iptablesChainPrefix = "ANTREA-POL"
+)
+
+/* NodeNetworkPolicy datapath implementation using iptables/ip6tables involves four components:
+1. Core iptables rule:
+   - Added to ANTREA-POL-INGRESS-RULES (ingress) or ANTREA-POL-EGRESS-RULES (egress). Note that, all ingress rules are
+     added to ANTREA-POL-INGRESS-RULES and all egress rules are added to ANTREA-POL-EGRESS-RULES.
+   - Matches an ipset created for the NodeNetworkPolicy rule as source (ingress) or destination (egress).
+   - Targets a service chain created for the NodeNetworkPolicy rule.
+2. Service iptables chain:
+   - Created for the NodeNetworkPolicy rule to integrate service iptables rules.
+3. Service iptables rules:
+   - Added to the service chain created for the NodeNetworkPolicy rule.
+   - Constructed from the services of the NodeNetworkPolicy rule.
+4. From/To ipSet:
+   - Created for the NodeNetworkPolicy rule, containing all source IP addresses (ingress) or destination IP addresses (egress).
+
+Assuming three ingress NodeNetworkPolicy rules with IDs 1111, 2222, and 3333, prioritized in descending order.
+Core iptables rules organized by priorities in ANTREA-POL-INGRESS-RULES like the following. Every ipset name consists of
+prefix "ANTREA-POL", rule ID and IP protocol version. The name the target chain consists of prefix "ANTREA-POL" and rule ID.
+
+```
+  ANTREA-POL-INGRESS-RULES
+  -A ANTREA-POL-INGRESS-RULES -m set --match-set ANTREA-POL-1111-4 src -j ANTREA-POL-1111 -m comment --comment "Antrea: for rule 1111, policy AntreaClusterNetworkPolicy:name1"
+  -A ANTREA-POL-INGRESS-RULES -m set --match-set ANTREA-POL-2222-4 src -j ANTREA-POL-2222 -m comment --comment "Antrea: for rule 2222, policy AntreaClusterNetworkPolicy:name2"
+  -A ANTREA-POL-INGRESS-RULES -m set --match-set ANTREA-POL-2222-4 src -j ANTREA-POL-2222 -m comment --comment "Antrea: for rule 3333, policy AntreaClusterNetworkPolicy:name3"
+```
+
+These are the service iptables chain and service iptables rules for the NodeNetworkPolicy rule with IDs 1111, 2222 and 3333.
+
+```
+ANTREA-POL-1111
+-A ANTREA-POL-1111 -j ACCEPT -p tcp --dport 80
+-A ANTREA-POL-1111 -j ACCEPT -p tcp --dport 443
+
+ANTREA-POL-2222
+-A ANTREA-POL-2222 -j ACCEPT -p tcp --dport 8080
+
+ANTREA-POL-3333
+-A ANTREA-POL-3333 -j ACCEPT -p tcp --dport 22
+```
+*/
+
+// ruleKey is a struct to identify a core iptables rule.
+type ruleKey struct {
+	priority types.Priority
+	ruleID   string
+}
+
+type iptRulesCache struct {
+	keyList *list.List        // keyList is a linked list that stores the `ruleKeys` of NodeNetworkPolicy rules in order.
+	rules   map[string]string // rules is a mapping from all NodeNetworkPolicy rule IDs to their core iptables rules.
```

Comment:
The data struct could be more concentrated and readable.
Given it needs to iterate all rules once anyway when adding/deleting a rule, having a mapping for ruleID to rule content is unnecessary; then you could just create a struct for a rule:
```
type iptRule struct {
        ruleID string
        priority *types.Priority
        rule string
}
````
then maintain rules sorted by priority.

Reply:
Done